build.sc

```diff
@@ -729,8 +729,8 @@ object contrib extends Module {
   }
 
   object buildinfo extends ContribModule {
-    def compileModuleDeps = Seq(scalalib, scalajslib, scalanativelib)
-    def testModuleDeps = super.testModuleDeps ++ Seq(scalalib, scalajslib, scalanativelib)
+    def moduleDeps = Seq(scalalib, scalajslib, scalanativelib)
```

Comment:
This means, if somebody includes this module into his project, it will also transitively include the Mill dependendencies. And if the versions do not match exactly, we end up with two different versions in the classpath.

Reply:
But as a contrib module, wouldn't the versions always line up? Since contrib modules are co-versioned with Mill

Reply:
Yeah, if used as we suggest. But I've seen also other constructs in the wild. And sometimes, these are even valid. We have no guard against it. The error are typically some strange compiler error messages and aren't helpful.

Reply:
Got it. I'll open a follow up to move back to `compileIvyDeps`

Reply:
> And if the versions do not match exactly, we end up with two different versions in the classpath.

Doesn't the version get evicted to the newer one?

Reply:
Now, that we control the whole build process of `build.sc` scripts with Mill, we could probably enhance our dependency management. E.g. we could apply some `forceVersion` to all Mill dependencies and such things.

Reply:
Yeah that works too. Or just skip them entirely during resolution (is that possible???) since they'll all be bundled anyway

Reply:
> Doesn't the version get evicted to the newer one?

It wouldn't, because the version bundled in the Mill assembly isnt resolved through ivy/maven and thus can't be evicted nor can it evict the other

Reply:
> > Doesn't the version get evicted to the newer one?
> 
> It wouldn't, because the version bundled in the Mill assembly isnt resolved through ivy/maven and thus can't be evicted nor can it evict the other

Also, in case we include a newer transitive Mill, it could still fail.

Maybe, we should try to resolve them from coursier and reduce the size of our bundle further. If we only come with our runner and coursier, we could load everything else via coursier on demand.

Reply:
Opened a PR to go back to the earlier behavior via https://github.com/com-lihaoyi/mill/pull/2562

The non-transitive nature of `compileModuleDeps` is definitely unfortunate; it means that anyone who depends on the `buildInfo` plugin needs to manually depend on `scalajslib` and `scalanativelib` as well. This isn't a problem for user code, but can be an annoyance for plugins.

I think the status quo is probably fine for 0.11.0. After that, we can consider other options, e.g. making use of `exclude` to allow normal dependencies while automatically removing the Mill jars from dependency resolution.

Shrinking the bundle is possible. With the current bootstrapping mechanism, we would need to keep at least `scalalib`, since that's required for `MillBootstrapRootModule`. With an alternative bootstrapping mechanism, we could cut it further, though it's unclear whether that's worth the complexity: for small size, the current bootstrap script already wins. We can leave that debate for later